working/macros/api/src/protocol/isolate_mirror_executor.dart

```diff
@@ -0,0 +1,131 @@
+/// a bridge between that isolate and the language frontends.
+class IsolateMirrorMacroExecutor implements MacroExecutor {
+  /// The actual isolate doing macro loading and execution.
+  final Isolate _macroIsolate;
+
+  /// The channel used to send requests to the [_macroIsolate].
+  final SendPort _sendPort;
+
+  /// The stream of responses from the [_macroIsolate].
+  final Stream<GenericResponse> _responseStream;
+
+  /// A map of response completers by request id.
+  final _responseCompleters = <int, Completer<GenericResponse>>{};
+
+  /// A function that should be invoked when shutting down this executor
+  /// to perform any necessary cleanup.
+  final void Function() _onClose;
+
+  IsolateMirrorMacroExecutor._(
+      this._macroIsolate, this._sendPort, this._responseStream, this._onClose) {
+    _responseStream.listen((event) {
+      var completer = _responseCompleters.remove(event.requestId);
+      if (completer == null) {
+        throw StateError(
+            'Got a response for an unrecognized request id ${event.requestId}');
+      }
+      completer.complete(event);
+    });
+  }
+
+  /// Initialize an [IsolateMirrorMacroExecutor] and return it once ready.
+  ///
+  /// Spawns the macro isolate and sets up a communication channel.
+  static Future<MacroExecutor> start() async {
+    var receivePort = ReceivePort();
+    var sendPortCompleter = Completer<SendPort>();
+    var responseStreamController =
+        StreamController<GenericResponse>(sync: true);
+    receivePort.listen((message) {
+      if (!sendPortCompleter.isCompleted) {
+        sendPortCompleter.complete(message as SendPort);
+      } else {
+        responseStreamController.add(message as GenericResponse);
+      }
+    }).onDone(responseStreamController.close);
+    var macroIsolate = await Isolate.spawn(spawn, receivePort.sendPort);
+
+    return IsolateMirrorMacroExecutor._(
+        macroIsolate,
+        await sendPortCompleter.future,
+        responseStreamController.stream,
+        receivePort.close);
+  }
+
+  @override
+  Future<String> buildAugmentationLibrary(
+      Iterable<MacroExecutionResult> macroResults) {
+    // TODO: implement buildAugmentationLibrary
+    throw UnimplementedError();
+  }
+
+  @override
+  void close() {
+    _onClose();
+    _macroIsolate.kill();
+  }
+
+  @override
+  Future<MacroExecutionResult> executeDeclarationsPhase(
+      MacroInstanceIdentifier macro,
+      Declaration declaration,
+      TypeComparator typeComparator,
+      ClassIntrospector classIntrospector) {
+    // TODO: implement executeDeclarationsPhase
+    throw UnimplementedError();
+  }
+
+  @override
+  Future<MacroExecutionResult> executeDefinitionsPhase(
+          MacroInstanceIdentifier macro,
+          Declaration declaration,
+          TypeComparator typeComparator,
+          ClassIntrospector classIntrospector,
+          TypeIntrospector typeIntrospector) =>
+      _sendRequest(ExecuteDefinitionsPhaseRequest(macro, declaration,
```

Comment:
We cannot send `TypeComparator` for example to another isolate, can we? It potentially a large data structure that references a big portion of analyzer or CFE. We probably have to marshal requests back and forth.

Reply:
Correct - a more real implementation would I think have its own internal rpcs to handle this type of introspection - which would delegate on the `executor` end of things to these objects. In fact in my actual original prototype that is exactly what it does.

This is just a starting point, more illustrative of how things are expected to work together than anything else at the moment.